persistence_legacy.go

```diff
@@ -183,61 +247,41 @@ func (db sqlitePersistence) MessageByChatID(chatID string, currCursor string, li
 	return result, newCursor, nil
 }
 
-func (db sqlitePersistence) MessagesFrom(from []byte) ([]*Message, error) {
-	allFields := db.tableUserMessagesLegacyAllFields()
-	rows, err := db.db.Query(
-		fmt.Sprintf(`
-			SELECT
-				%s
-			FROM
-				user_messages_legacy
-			WHERE
-				source = ?
-		`, allFields),
-		from,
+func (db sqlitePersistence) SaveMessagesLegacy(messages []*Message) error {
+	var (
+		tx   *sql.Tx
+		stmt *sql.Stmt
+		err  error
 	)
+	tx, err = db.db.BeginTx(context.Background(), &sql.TxOptions{})
 	if err != nil {
-		return nil, err
+		return err
 	}
-	defer rows.Close()
+	allFields := db.tableUserMessagesLegacyAllFields()
+	valuesVector := strings.Repeat("?, ", db.tableUserMessagesLegacyAllFieldsCount()-1) + "?"
 
-	var result []*Message
-	for rows.Next() {
-		var message Message
-		if err := db.tableUserMessagesLegacyScanAllFields(rows, &message); err != nil {
-			return nil, err
-		}
-		result = append(result, &message)
-	}
-	return result, nil
-}
+	query := fmt.Sprintf(`INSERT INTO user_messages_legacy(%s) VALUES (%s)`, allFields, valuesVector)
 
-func (db sqlitePersistence) UnseenMessageIDs() ([][]byte, error) {
-	rows, err := db.db.Query(`SELECT id FROM user_messages_legacy WHERE seen = 0`)
+	stmt, err = tx.Prepare(query)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	defer rows.Close()
+	defer func() {
```

Comment:
This should be right after checking error for `tx, err = ...`. I think that's the place where a tx is created. If `tx.Prepare()` returns an error, the tx will be never completed.